cluster/encryption.go

```diff
@@ -0,0 +1,152 @@
+package cluster
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/rancher/rke/k8s"
+	"github.com/rancher/rke/services"
+	"github.com/rancher/rke/templates"
+	"github.com/sirupsen/logrus"
+)
+
+// DeploySecretsEncryptionConfig -  Deploy encryption.yaml to ControlPlane hosts.
+func (c *Cluster) DeploySecretsEncryptionConfig(ctx context.Context, currentCluster *Cluster) error {
+	if currentCluster == nil {
+		currentCluster = &Cluster{}
+	}
+
+	if c.Services.KubeAPI.SecretsEncryptionConfig == currentCluster.Services.KubeAPI.SecretsEncryptionConfig {
+		logrus.Debug("[controlplane] Secrets encryption config matches state. No changes needed.")
+		return nil
+	}
+
+	for _, host := range c.ControlPlaneHosts {
+		err := c.WriteFile(ctx, host, "controlplane", "/etc/kubernetes/encryption.yaml", "0600", c.Services.KubeAPI.SecretsEncryptionConfig)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ReconcileSecretsEncryptionConfig - Restart controlplane and replace all secrets in the cluster when the encryption config changes.
+func (c *Cluster) ReconcileSecretsEncryptionConfig(ctx context.Context, currentCluster *Cluster) error {
+	if currentCluster != nil {
+		if c.Services.KubeAPI.SecretsEncryptionConfig == currentCluster.Services.KubeAPI.SecretsEncryptionConfig {
+			logrus.Debug("[reconcile] Secrets encryption config matches. No changes needed.")
+			return nil
+		}
+
+		logrus.Info("[reconcile] New secrets encryption config. Updating existing secrets to apply encryption.")
+
+		err := services.RestartControlPlane(ctx, c.ControlPlaneHosts)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := k8s.NewClient(c.LocalKubeConfigPath, c.K8sWrapTransport)
+		if err != nil {
+			return fmt.Errorf("Failed to initialize new kubernetes client: %v", err)
+		}
+
+		secrets, err := k8s.GetAllSecrets(kubeClient)
+		if err != nil {
+			return err
+		}
+
+		for _, secret := range secrets {
```

Comment:
This should be done in a concurrent worker loop like this: https://github.com/rancher/rke/blob/master/cluster/cluster.go#L366-L384

Reply:
If you have potentially 1000s of secrets, should they be done concurrently. Is there a potential to hit rate limits? Enabling EventRateLimiter is on the list of recommended security options.

Reply:
I'll admit that I don't have a lot of experience with concurrency. I'm confused by the code you reference.  

It looks to me like each thread is going to iterate through the entire hostQueue and its going to repeat the creation of a thread and iterating through the hostQueue SyncWorkers(10) times.

What am I missing here?